### PR TITLE
Adds the Organ Extractor

### PR DESCRIPTION
**Pull request by Qwertytoforty**
## What Does This PR Do

Inspired by VG's organ extractor, and expanded upon.

An item traitors can buy for 20 TC. On a horizontal / dead humanoid, they can select an organ in the mob, and begin drilling into the mob for their organ. After 12 seconds, they will extract the organ from the target, and store it inside the extractors jar.
They can keep the organ in the jar where it will not decay, remove it from the jar, or use the extractor in hand, or on another mob to forcefully inject the organ into them.

You can scredriver it to adjust the organ configuration if it is an arm, or emag it to emag the internal organ.

ONLY WORKS ON MOBS THAT HAVE OR HAD A CLIENT.

You can not drill into a monkey you stuffed full of combat implants.
You could stuff an actual person with combat implants then use the tool, but at that point you are at a greater risk of getting caught. And if the person you are implanting in is an ally, well you could have gotten them to implant you.

Dead bodies that have been debrained or cloned still worked, so raiding the morgue is a valid option.

Can fit on a medical belt, not that doctors will often buy it, but more because it is sorta a medical tool, and frankly funny, you could just put it in a box anyway.

## Why It's Good For The Game

A traitor item with many potential uses. An expensive alternative to debraining, without needing a table, but certain to work.
A way to remove implants or other valuable organs from someone. Be it CNS, a toolset implant, or a pulse demon heart, it can be extracted.
Why not debrain that dead EOC from earlier that helped you, and inject their brain in your target? Win win!
Removing xenomorph larva from allies, and putting it into a target. Or just removing it period.
Could be useful when the traitor cybernetics are merged, if your target is a traitor, or for raiding  the morgue.

## Images of changes

Image of some organs in the tank.
image

## Testing

Drilled into a unathi and grey a bunch. Failed to drill simple mobs, can't drill when drilling, can't drill and move, etc, you don't delete brains with it, don't die when implanting a heart into yourself.

## Changelog
:cl:
add: Adds the Organ Extractor. For 20 TC, you can get a useful tool for removing organs or cybernetics from restrained or dead people, that can also be used to implant the removed organs into yourself.
/:cl:

### CODE REVIEW

File: code/modules/surgery/organs/organ_extractor.dm [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+	C.apply_damage(5, BRUTE, drilled_organ)
+	playsound(get_turf(C), 'sound/weapons/circsawhit.ogg', 50, TRUE)
+	if(user_is_target)
+		if(!do_after_once(C, 7 SECONDS, target = C))
+			to_chat(user, "<span class='warning'>ERROR: Process inturupted!</span>")
+			in_use = FALSE
+			return
+	else
+		if(!do_after_once(C, 12 SECONDS, target = C))
+			to_chat(user, "<span class='warning'>ERROR: Process inturupted!</span>")
```
Comment (Henri215) [inline code/modules/surgery/organs/organ_extractor.dm:141]:
```suggestion
			to_chat(user, "<span class='warning'>ERROR: Process interrupted!</span>")
			in_use = FALSE
			return
	else
		if(!do_after_once(C, 12 SECONDS, target = C))
			to_chat(user, "<span class='warning'>ERROR: Process interrupted!</span>")
```

File: code/modules/surgery/organs/organ_extractor.dm [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+		in_use = FALSE
+		return
+	var/obj/item/organ/internal/internal_organ = chosen_organ
+	var/drilled_organ = internal_organ.parent_organ
+	user.visible_message("<span class='danger'>[user] activates [src] and begins to drill into [C]!</span>", "<span class='warning'>You level the extractor at [M] and hold down the trigger.</span>")
+	to_chat(C, "<span class='danger'>You feel a lot of pain as [user] drills into your [drilled_organ]!</span>")
+	playsound(get_turf(user), 'sound/weapons/circsawhit.ogg', 75, TRUE)
+	C.apply_damage(15, BRUTE, drilled_organ)
+	if(!do_after_once(user, 12 SECONDS, target = C))// Slightly longer than stamina crit, at least cuff and buckle them to a pipe or something
+		to_chat(user, "<span class='warning'>ERROR: Process inturupted!</span>")
```
Comment (Henri215) [inline code/modules/surgery/organs/organ_extractor.dm:84]:
```suggestion
		to_chat(user, "<span class='warning'>ERROR: Process interrupted!</span>")
```

File: code/modules/surgery/organs/organ_extractor.dm [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+	internal_organ.remove(C)
+	internal_organ.forceMove(src)
+	storedorgan = internal_organ
+	storedorgan.rejuvenate() //Organ gets dumped into the internal mito tank, heals it up. And nanites for robotic organs, I guess.
+	in_use = FALSE
+	var/xx = internal_organ.pixel_x
+	var/yy = internal_organ.pixel_y
+	internal_organ.pixel_x = 2
+	internal_organ.pixel_y = -2
+	var/image/img = image("icon"=internal_organ, "layer"=FLOAT_LAYER)
```
Comment (Henri215) [inline code/modules/surgery/organs/organ_extractor.dm:103]:
```suggestion
	var/image/img = image("icon" = internal_organ, "layer" = FLOAT_LAYER)
```

File: code/modules/surgery/organs/organ_extractor.dm [hunk 4]
```diff
@@ -0,0 +1,156 @@
+/obj/item/organ_extractor
+	name = "organ extractor"
+	desc = "A device that can remove organs from a target, and store them inside. Stored organs can be implanted into the user. Synthesizes chemicals to keep the organ fresh."
+	icon = 'icons/obj/surgery.dmi'
+	icon_state = "codersprite"
```
Comment (Henri215) [inline code/modules/surgery/organs/organ_extractor.dm:5]: Dont forget to change the sprite name.

Comment (Qwertytoforty) [reply code/modules/surgery/organs/organ_extractor.dm:5]: done

**Comment by mariadkgdmgdmi** [comment]
autosurgeon for 20 tc

**Comment by Qwertytoforty** [comment]
> autosurgeon for 20 tc

You can not directly implant an organ into the extractor, nor can you put an organ in a dead body to extract it.

File: code/modules/surgery/organs/organ_extractor.dm [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+	user.visible_message("<span class='danger'>[user] removes [internal_organ] from [C]!</span>", "<span class='warning'>You remove [internal_organ] from [C] as it gets sucked into [src]'s internal container!</span>")
+	playsound(get_turf(user), 'sound/weapons/circsawhit.ogg', 75, TRUE)
+	C.apply_damage(10, BRUTE, drilled_organ)
+	internal_organ.remove(C)
+	internal_organ.forceMove(src)
+	storedorgan = internal_organ
+	storedorgan.rejuvenate() //Organ gets dumped into the internal mito tank, heals it up. And nanites for robotic organs, I guess.
+	in_use = FALSE
+	var/xx = internal_organ.pixel_x
+	var/yy = internal_organ.pixel_y
```
Comment (DGamerL) [inline code/modules/surgery/organs/organ_extractor.dm:100]: Maybe rename these to `organ_x` and `organ_y`?

File: code/modules/surgery/organs/organ_extractor.dm [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+	if(user == our_target)
+		user_is_target = TRUE
+	if(istype(storedorgan, /obj/item/organ/internal/brain) && user_is_target)
+		to_chat(user, "<span class='warning'>It would be really stupid to replace your brain with another one.</span>")
+		return
+	if(!iscarbon(our_target))
+		return
+	var/mob/living/carbon/C = our_target
+	in_use = TRUE
+	user.visible_message("<span class='danger'>[user] activates [src] and begins to drill into [C]!</span>", "<span class='warning'>You level the extractor at [C] and hold down the trigger.</span>")
```
Comment (DGamerL) [inline code/modules/surgery/organs/organ_extractor.dm:130]:
```suggestion
	user.visible_message("<span class='danger'>[user] activates [src] and begins to drill into [C]!</span>", "<span class='warning'>You level the extractor at [user_is_target ? "yourself" : C] and hold down the trigger.</span>")
```

**Review by S34NW — APPROVED** [review]

**Pull request merged**